File: tev_script/runtime_v3.py
from __future__ import annotations

from collections import deque
from copy import deepcopy
from dataclasses import dataclass
from fractions import Fraction
import re
from typing import Any, Callable, Mapping

from .diagnostics import TevScriptError
from .ir_v3_validation import validate_program_ir_v3
from .ir_v3_values import (
    RecordValueV3,
    TypeTableV3,
    VariantValueV3,
    decode_v3_value,
    encode_v3_value,
    v3_values_equal,
)

CapabilityV3 = Callable[..., Any]
_STABLE_ID = re.compile(r"^[A-Za-z_][A-Za-z0-9_.:/-]*$")
_LOCAL_ID = re.compile(r"^[A-Za-z_][A-Za-z0-9_]*$")


@dataclass(frozen=True, slots=True)
class EmittedEventV3:
    entity_id: str
    event_id: str
    argument_types: tuple[str, ...]
    arguments: tuple[Any, ...]

    def canonical_arguments(self, table: TypeTableV3) -> tuple[Any, ...]:
        return tuple(
            encode_v3_value(type_id, value, table, context=f"event {self.event_id}")
            for type_id, value in zip(self.argument_types, self.arguments, strict=True)
        )


@dataclass(slots=True)
class EntityRuntimeV3:
    entity_id: str
    state: dict[str, Any]
    state_types: dict[str, str]
    handlers: dict[str, dict[str, Any]]
    emitted_event_types: dict[str, tuple[str, ...]]


class ScriptRuntimeV3:
    def __init__(
        self,
        ir: Mapping[str, Any],
        capabilities: Mapping[str, CapabilityV3] | None = None,
        *,
        expected_source_semantic_hash: str | None = None,
    ) -> None:
        self.ir = deepcopy(dict(ir))
        self.type_table = validate_program_ir_v3(
            self.ir,
            expected_source_semantic_hash=expected_source_semantic_hash,
        )
        self.capabilities = dict(capabilities or {})
        for capability_id in self.capabilities:
            if _STABLE_ID.fullmatch(capability_id) is None:
                raise TevScriptError(
                    "TEVS_IR_V3_CAPABILITY_BINDING_ID",
                    f"non-canonical capability binding id {capability_id!r}",
                )
        self.entities: dict[str, EntityRuntimeV3] = {}
        self.emitted: list[EmittedEventV3] = []
        for raw_entity in self.ir["entities"]:
            state: dict[str, Any] = {}
            state_types: dict[str, str] = {}
            for raw_state in raw_entity["states"]:
                name = str(raw_state["name"])
                type_id = str(raw_state["type"])
                state[name] = decode_v3_value(
                    type_id,
                    raw_state["initial"],
                    self.type_table,
                    context=f"state {raw_entity['entity_id']}.{name}",
                )
                state_types[name] = type_id
            handlers = {
                str(handler["event_id"]): handler
                for handler in raw_entity["handlers"]
            }
            event_types = {
                str(event["event_id"]): tuple(str(value) for value in event["parameters"])
                for event in raw_entity["emitted_events"]
            }
            entity = EntityRuntimeV3(
                str(raw_entity["entity_id"]),
                state,
                state_types,
                handlers,
                event_types,
            )
            self.entities[entity.entity_id] = entity

    @property
    def source_semantic_hash(self) -> str:
        return str(self.ir["source_semantic_hash"])

    @property
    def semantic_hash(self) -> str:
        return str(self.ir["semantic_hash"])

    def invoke(
        self,
        entity_id: str,
        event_id: str,
        *arguments: Any,
    ) -> tuple[EmittedEventV3, ...]:
        _require_local_id(entity_id, "entity")
        _require_local_id(event_id, "event")
        entity = self.entities.get(entity_id)
        if entity is None:
            raise TevScriptError("TEVS_IR_V3_ENTITY_UNKNOWN", f"unknown entity {entity_id!r}")

        start = len(self.emitted)
        queue: deque[tuple[str, tuple[Any, ...]]] = deque()
        queue.append((event_id, tuple(arguments)))
        maximum = int(self.ir["boundary"]["maximum_event_chain"])
        processed = 0

        while queue:
            processed += 1
            if processed > maximum:
                raise TevScriptError(
                    "TEVS_IR_V3_EVENT_BUDGET",
                    f"event chain exceeds {maximum}",
                )
            current_event, raw_arguments = queue.popleft()
            handler = entity.handlers.get(current_event)
            if handler is None:
                # External/unhandled invocation has no declaration from which a
                # portable argument signature can be inferred. This matches the
                # historical local-handler event model: only emitted events are
                # recorded canonically.
                continue
            parameters = handler["parameters"]
            if len(parameters) != len(raw_arguments):
                raise TevScriptError(
                    "TEVS_IR_V3_EVENT_ARITY",
                    f"event {current_event!r} expects {len(parameters)} arguments",
                )
            parameter_values = {
                str(parameter["name"]): _coerce_runtime_v3(
                    str(parameter["type"]),
                    raw_arguments[index],
                    self.type_table,
                    context=f"invoke {entity_id}.{current_event}[{index}]",
                )
                for index, parameter in enumerate(parameters)
            }
            generated = self._execute_handler(entity, handler, parameter_values)
            for generated_event, generated_types, generated_arguments in generated:
                emitted = EmittedEventV3(
                    entity_id,
                    generated_event,
                    generated_types,
                    generated_arguments,
                )
                self.emitted.append(emitted)
                if generated_event in entity.handlers:
                    queue.append((generated_event, generated_arguments))
        return tuple(self.emitted[start:])

    def state(self, entity_id: str) -> dict[str, Any]:
        entity = self.entities.get(entity_id)
        if entity is None:
            raise KeyError(entity_id)
        return dict(entity.state)

    def canonical_state(self, entity_id: str) -> dict[str, Any]:
        entity = self.entities.get(entity_id)
        if entity is None:
            raise KeyError(entity_id)
        return {
            name: encode_v3_value(
                entity.state_types[name],
                entity.state[name],
                self.type_table,
                context=f"state {entity_id}.{name}",
            )
            for name in sorted(entity.state)
        }

    def _execute_handler(
        self,
        entity: EntityRuntimeV3,
        handler: Mapping[str, Any],
        parameters: Mapping[str, Any],
    ) -> list[tuple[str, tuple[str, ...], tuple[Any, ...]]]:
        instructions = list(handler["instructions"])
        budget = int(handler["instruction_budget"])
        stack: list[Any] = []
        locals_: dict[str, Any] = {}
        emitted: list[tuple[str, tuple[str, ...], tuple[Any, ...]]] = []
        pc = 0
        executed = 0

        while pc < len(instructions):
            executed += 1
            if executed > budget:
                raise TevScriptError(
                    "TEVS_IR_V3_INSTRUCTION_BUDGET",
                    f"handler {handler['event_id']!r} exceeded instruction budget {budget}",
                )
            instruction = instructions[pc]
            op = str(instruction["op"])

            if op == "CONST":
                stack.append(
                    decode_v3_value(
                        str(instruction["type"]),
                        instruction["value"],
                        self.type_table,
                        context=f"instruction {handler['event_id']}:{pc}",
                    )
                )
            elif op == "LOAD_STATE":
                stack.append(entity.state[str(instruction["name"])])
            elif op == "STORE_STATE":
                entity.state[str(instruction["name"])] = stack.pop()
            elif op == "LOAD_LOCAL":
                stack.append(locals_[str(instruction["name"])])
            elif op == "STORE_LOCAL":
                locals_[str(instruction["name"])] = stack.pop()
            elif op == "LOAD_PARAM":
                stack.append(parameters[str(instruction["name"])])
            elif op == "CONVERT_INT_TO_RAT":
                stack.append(Fraction(stack.pop(), 1))
            elif op == "UNARY":
                stack.append(_unary(str(instruction["operator"]), stack.pop()))
            elif op == "BINARY":
                right = stack.pop()
                left = stack.pop()
                stack.append(
                    _binary_v3(
                        str(instruction["operator"]),
                        str(instruction["left_type"]),
                        str(instruction["right_type"]),
                        left,
                        right,
                        self.type_table,
                    )
                )
            elif op == "CALL_PURE":
                args = _pop_arguments(stack, int(instruction["argc"]))
                result = _call_pure(str(instruction["function_id"]), args)
                if str(instruction["return_type"]) != "Unit":
                    stack.append(result)
            elif op == "CALL_CAPABILITY":
                capability_id = str(instruction["capability_id"])
                capability = self.capabilities.get(capability_id)
                if capability is None:
                    raise TevScriptError(
                        "TEVS_IR_V3_CAPABILITY_MISSING",
                        f"capability {capability_id!r} is not bound",
                    )
                args = _pop_arguments(stack, int(instruction["argc"]))
                result = capability(*args)
                return_type = str(instruction["return_type"])
                if return_type != "Unit":
                    stack.append(
                        _coerce_runtime_v3(
                            return_type,
                            result,
                            self.type_table,
                            context=f"capability {capability_id} return",
                        )
                    )
            elif op == "EMIT_EVENT":
                argument_types = tuple(str(value) for value in instruction["argument_types"])
                args = tuple(_pop_arguments(stack, int(instruction["argc"])))
                # Values are already typed by the verified stack. Canonicalize and
                # decode once to guarantee host representation cannot leak.
                args = tuple(
                    _coerce_runtime_v3(
                        type_id,
                        value,
                        self.type_table,
                        context=f"emit {instruction['event_id']}[{index}]",
                    )
                    for index, (type_id, value) in enumerate(zip(argument_types, args, strict=True))
                )
                emitted.append((str(instruction["event_id"]), argument_types, args))
            elif op == "MAKE_RECORD":
                type_id = str(instruction["type"])
                descriptor = self.type_table.require(type_id)
                fields = tuple(str(name) for name in instruction["fields"])
                values = _pop_arguments(stack, len(fields))
                by_name = dict(zip(fields, values, strict=True))
                record = RecordValueV3(
                    type_id,
                    tuple((name, by_name[name]) for name, _field_type in descriptor.fields),
                )
                # Encoding validates nested semantic representations and returns a
                # canonical representation; decode normalizes any host containers.
                stack.append(
                    decode_v3_value(
                        type_id,
                        encode_v3_value(type_id, record, self.type_table),
                        self.type_table,
                    )
                )
            elif op == "LOAD_FIELD":
                record = stack.pop()
                if not isinstance(record, RecordValueV3):
                    raise TevScriptError("TEVS_IR_V3_RUNTIME_RECORD", "LOAD_FIELD received non-record value")
                stack.append(record.field(str(instruction["field"])))
            elif op == "MAKE_VARIANT":
                type_id = str(instruction["type"])
                variant = str(instruction["variant"])
                argc = int(instruction["argc"])
                if argc == 0:
                    value = VariantValueV3(type_id, variant)
                else:
                    value = VariantValueV3(type_id, variant, stack.pop())
                stack.append(
                    decode_v3_value(
                        type_id,
                        encode_v3_value(type_id, value, self.type_table),
                        self.type_table,
                    )
                )
            elif op == "TEST_VARIANT":
                value = stack.pop()
                if not isinstance(value, VariantValueV3):
                    raise TevScriptError("TEVS_IR_V3_RUNTIME_VARIANT", "TEST_VARIANT received non-variant value")
                stack.append(value.variant == str(instruction["variant"]))
            elif op == "LOAD_VARIANT_PAYLOAD":
                value = stack.pop()
                variant = str(instruction["variant"])
                if not isinstance(value, VariantValueV3) or value.variant != variant or not value.has_payload:
                    raise TevScriptError(
                        "TEVS_IR_V3_VARIANT_UNWRAP",
                        f"cannot load payload for variant {variant!r}",
                    )
                stack.append(value.payload)
            elif op == "JUMP_IF_FALSE":
                condition = stack.pop()
                if condition is not True:
                    pc = int(instruction["target"])
                    continue
            elif op == "JUMP":
                pc = int(instruction["target"])
                continue
            elif op == "RETURN":
                break
            else:
                raise TevScriptError("TEVS_IR_V3_OPCODE", f"unknown V3 opcode {op!r}")
            pc += 1

        if stack:
            raise TevScriptError(
                "TEVS_IR_V3_RUNTIME_STACK_LEAK",
                f"handler {handler['event_id']!r} left {len(stack)} values on stack",
            )
        return emitted


def _coerce_runtime_v3(type_id: str, value: Any, table: TypeTableV3, *, context: str) -> Any:
    descriptor = table.require(type_id, context=context)
    if descriptor.kind == "primitive":
        if type_id == "Rat":
            if isinstance(value, Fraction):
                return value
            if isinstance(value, bool):
                raise TevScriptError("TEVS_IR_V3_CAPABILITY_COERCION", f"{context}: bool is not Rat")
            if isinstance(value, int):
                return Fraction(value, 1)
        elif type_id == "Int":
            if isinstance(value, bool) or not isinstance(value, int):
                raise TevScriptError("TEVS_IR_V3_CAPABILITY_COERCION", f"{context}: expected Int")
            return value
        elif type_id == "Bool":
            if not isinstance(value, bool):
                raise TevScriptError("TEVS_IR_V3_CAPABILITY_COERCION", f"{context}: expected Bool")
            return value
        elif type_id == "Text":
            if not isinstance(value, str):
                raise TevScriptError("TEVS_IR_V3_CAPABILITY_COERCION", f"{context}: expected Text")
            return value
        elif type_id in {"Vec2", "Vec3"}:
            expected = 2 if type_id == "Vec2" else 3
            if not isinstance(value, (tuple, list)) or len(value) != expected:
                raise TevScriptError("TEVS_IR_V3_CAPABILITY_COERCION", f"{context}: expected {type_id}")
            return tuple(_coerce_runtime_v3("Rat", item, table, context=context) for item in value)
        raise AssertionError(type_id)

    if isinstance(value, (RecordValueV3, VariantValueV3)):
        raw = encode_v3_value(type_id, value, table, context=context)
        return decode_v3_value(type_id, raw, table, context=context)
    if isinstance(value, dict):
        return decode_v3_value(type_id, value, table, context=context)
    raise TevScriptError(
        "TEVS_IR_V3_CAPABILITY_COERCION",
        f"{context}: composite value must be a canonical encoded value or V3 semantic value",
    )


def _binary_v3(
    operator: str,
    left_type: str,
    right_type: str,
    left: Any,
    right: Any,
    table: TypeTableV3,
) -> Any:
    if operator == "AND":
        return bool(left and right)
    if operator == "OR":
        return bool(left or right)
    if operator in {"EQEQ", "NE"}:
        if {left_type, right_type} == {"Int", "Rat"}:
            equal = Fraction(left) == Fraction(right)
        elif left_type == right_type:
            equal = v3_values_equal(left_type, left, right, table)
        else:
            raise TevScriptError("TEVS_IR_V3_RUNTIME_BINARY", "invalid equality types")
        return equal if operator == "EQEQ" else not equal
    if operator == "LT":
        return left < right
    if operator == "LE":
        return left <= right
    if operator == "GT":
        return left > right
    if operator == "GE":
        return left >= right
    if operator == "PLUS":
        return _vector_or_scalar(left, right, lambda a, b: a + b)
    if operator == "MINUS":
        return _vector_or_scalar(left, right, lambda a, b: a - b)
    if operator == "STAR":
        if isinstance(left, tuple) and not isinstance(right, tuple):
            return tuple(item * right for item in left)
        if isinstance(right, tuple) and not isinstance(left, tuple):
            return tuple(left * item for item in right)
        return left * right
    if operator == "SLASH":
        if right == 0:
            raise TevScriptError("TEVS_IR_V3_DIVIDE_ZERO", "division by zero")
        if isinstance(left, tuple):
            return tuple(Fraction(item) / Fraction(right) for item in left)
        return Fraction(left) / Fraction(right)
    raise TevScriptError("TEVS_IR_V3_RUNTIME_BINARY", f"unknown binary operator {operator!r}")


def _unary(operator: str, value: Any) -> Any:
    if operator == "NOT":
        return not value
    if operator == "MINUS":
        return -value
    raise TevScriptError("TEVS_IR_V3_RUNTIME_UNARY", f"unknown unary operator {operator!r}")


def _vector_or_scalar(left: Any, right: Any, operation: Callable[[Any, Any], Any]) -> Any:
    if isinstance(left, tuple) or isinstance(right, tuple):
        if not isinstance(left, tuple) or not isinstance(right, tuple):
            raise TevScriptError("TEVS_IR_V3_VECTOR_OPERAND", "vector plus/minus requires two vectors")
        return tuple(operation(a, b) for a, b in zip(left, right, strict=True))
    return operation(left, right)


def _call_pure(function_id: str, arguments: list[Any]) -> Any:
    if function_id == "vec2":
        return tuple(Fraction(item) for item in arguments)
    if function_id == "vec3":
        return tuple(Fraction(item) for item in arguments)
    if function_id == "max":
        return max(arguments)
    if function_id == "min":
        return min(arguments)
    raise TevScriptError("TEVS_IR_V3_PURE_FUNCTION", f"unknown pure intrinsic {function_id!r}")


def _pop_arguments(stack: list[Any], count: int) -> list[Any]:
    if count == 0:
        return []
    if len(stack) < count:
        raise TevScriptError("TEVS_IR_V3_STACK_UNDERFLOW", "not enough values for operation")
    values = stack[-count:]
    del stack[-count:]
    return values


def _require_local_id(value: str, kind: str) -> None:
    if _LOCAL_ID.fullmatch(value) is None:
        raise TevScriptError("TEVS_IR_V3_INVOCATION_ID", f"non-canonical {kind} id {value!r}")

File: javascript/src/runtime-v3.mjs
import { compareValues, coerceRuntime, Rational, TevScriptError } from "./values.mjs";
import { validateProgramIrV3 } from "./ir-v3-validation.mjs";
import {
  RecordValueV3,
  VariantValueV3,
  decodeV3Value,
  encodeV3Value,
  v3ValuesEqual,
} from "./ir-v3-values.mjs";

const STABLE_ID = /^[A-Za-z_][A-Za-z0-9_.:/-]*$/;
const LOCAL_ID = /^[A-Za-z_][A-Za-z0-9_]*$/;

function requireLocalId(value, kind) {
  if (typeof value !== "string" || !LOCAL_ID.test(value)) {
    throw new TevScriptError("TEVS_IR_V3_INVOCATION_ID", `non-canonical ${kind} id ${JSON.stringify(value)}`);
  }
  return value;
}

function requireBindingId(value) {
  if (typeof value !== "string" || !STABLE_ID.test(value)) {
    throw new TevScriptError("TEVS_IR_V3_CAPABILITY_BINDING_ID", `non-canonical capability binding id ${JSON.stringify(value)}`);
  }
  return value;
}

function popArguments(stack, count) {
  if (count === 0) return [];
  if (stack.length < count) {
    throw new TevScriptError("TEVS_IR_V3_STACK_UNDERFLOW", "not enough values for operation");
  }
  return stack.splice(stack.length - count, count);
}

function add(left, right) {
  if (left instanceof Rational || right instanceof Rational) return Rational.from(left).add(right);
  return left + right;
}
function subtract(left, right) {
  if (left instanceof Rational || right instanceof Rational) return Rational.from(left).sub(right);
  return left - right;
}
function multiply(left, right) {
  if (left instanceof Rational || right instanceof Rational) return Rational.from(left).mul(right);
  return left * right;
}
function vectorOrScalar(left, right, operation) {
  if (Array.isArray(left) || Array.isArray(right)) {
    if (!Array.isArray(left) || !Array.isArray(right) || left.length !== right.length) {
      throw new TevScriptError("TEVS_IR_V3_VECTOR_OPERAND", "vector plus/minus requires equal vectors");
    }
    return left.map((value, index) => operation(value, right[index]));
  }
  return operation(left, right);
}

function unary(operator, value) {
  if (operator === "NOT") return !value;
  if (operator === "MINUS") {
    if (value instanceof Rational) return value.neg();
    return -value;
  }
  throw new TevScriptError("TEVS_IR_V3_RUNTIME_UNARY", `unknown unary operator ${operator}`);
}

function binaryV3(operator, leftType, rightType, left, right, table) {
  if (operator === "AND") return Boolean(left && right);
  if (operator === "OR") return Boolean(left || right);
  if (operator === "EQEQ" || operator === "NE") {
    let equal;
    if ((leftType === "Int" && rightType === "Rat") || (leftType === "Rat" && rightType === "Int")) {
      equal = Rational.from(left).equals(Rational.from(right));
    } else if (leftType === rightType) {
      equal = v3ValuesEqual(leftType, left, right, table);
    } else {
      throw new TevScriptError("TEVS_IR_V3_RUNTIME_BINARY", "invalid equality types");
    }
    return operator === "EQEQ" ? equal : !equal;
  }
  if (operator === "LT") return compareValues(left, right) < 0;
  if (operator === "LE") return compareValues(left, right) <= 0;
  if (operator === "GT") return compareValues(left, right) > 0;
  if (operator === "GE") return compareValues(left, right) >= 0;
  if (operator === "PLUS") return vectorOrScalar(left, right, add);
  if (operator === "MINUS") return vectorOrScalar(left, right, subtract);
  if (operator === "STAR") {
    if (Array.isArray(left) && !Array.isArray(right)) return left.map((item) => multiply(item, right));
    if (Array.isArray(right) && !Array.isArray(left)) return right.map((item) => multiply(left, item));
    return multiply(left, right);
  }
  if (operator === "SLASH") {
    if (Array.isArray(left)) return left.map((item) => Rational.from(item).div(right));
    return Rational.from(left).div(right);
  }
  throw new TevScriptError("TEVS_IR_V3_RUNTIME_BINARY", `unknown binary operator ${operator}`);
}

function callPure(functionId, args) {
  if (functionId === "vec2" || functionId === "vec3") return args.map((item) => Rational.from(item));
  if (functionId === "max" || functionId === "min") {
    if (args.length !== 2) throw new TevScriptError("TEVS_IR_V3_PURE_ARITY", functionId);
    const comparison = compareValues(args[0], args[1]);
    return functionId === "max"
      ? comparison >= 0 ? args[0] : args[1]
      : comparison <= 0 ? args[0] : args[1];
  }
  throw new TevScriptError("TEVS_IR_V3_PURE_FUNCTION", `unknown pure intrinsic ${functionId}`);
}

function coerceRuntimeV3(typeId, value, table, context) {
  const descriptor = table.require(typeId, context);
  if (descriptor.kind === "primitive") {
    try {
      return coerceRuntime(value, typeId);
    } catch (error) {
      throw new TevScriptError(
        "TEVS_IR_V3_CAPABILITY_COERCION",
        `${context}: ${error.message}`,
      );
    }
  }
  if (value instanceof RecordValueV3 || value instanceof VariantValueV3) {
    const raw = encodeV3Value(typeId, value, table, { context });
    return decodeV3Value(typeId, raw, table, { context });
  }
  if (value !== null && typeof value === "object" && !Array.isArray(value)) {
    return decodeV3Value(typeId, value, table, { context });
  }
  throw new TevScriptError(
    "TEVS_IR_V3_CAPABILITY_COERCION",
    `${context}: composite value must be canonical encoded value or V3 semantic value`,
  );
}

export class ScriptRuntimeV3 {
  constructor(ir, capabilities = {}, { expectedSourceSemanticHash = null } = {}) {
    this.ir = structuredClone(ir);
    this.typeTable = validateProgramIrV3(this.ir, { expectedSourceSemanticHash });
    this.capabilities = { ...capabilities };
    for (const capabilityId of Reflect.ownKeys(this.capabilities)) requireBindingId(capabilityId);
    this.entities = new Map();
    this.emitted = [];

    for (const rawEntity of this.ir.entities) {
      const state = new Map();
      const stateTypes = new Map();
      for (const rawState of rawEntity.states) {
        state.set(
          rawState.name,
          decodeV3Value(rawState.type, rawState.initial, this.typeTable, {
            context: `state ${rawEntity.entity_id}.${rawState.name}`,
          }),
        );
        stateTypes.set(rawState.name, rawState.type);
      }
      const handlers = new Map(rawEntity.handlers.map((handler) => [handler.event_id, handler]));
      const emittedEventTypes = new Map(
        rawEntity.emitted_events.map((event) => [event.event_id, [...event.parameters]]),
      );
      this.entities.set(rawEntity.entity_id, {
        entityId: rawEntity.entity_id,
        state,
        stateTypes,
        handlers,
        emittedEventTypes,
      });
    }
  }

  get sourceSemanticHash() { return this.ir.source_semantic_hash; }
  get semanticHash() { return this.ir.semantic_hash; }

  invoke(entityId, eventId, ...args) {
    entityId = requireLocalId(entityId, "entity");
    eventId = requireLocalId(eventId, "event");
    const entity = this.entities.get(entityId);
    if (!entity) throw new TevScriptError("TEVS_IR_V3_ENTITY_UNKNOWN", `unknown entity ${JSON.stringify(entityId)}`);
    const start = this.emitted.length;
    const queue = [[eventId, args]];
    const maximum = Number(this.ir.boundary.maximum_event_chain);
    let processed = 0;

    while (queue.length > 0) {
      processed += 1;
      if (processed > maximum) throw new TevScriptError("TEVS_IR_V3_EVENT_BUDGET", `event chain exceeds ${maximum}`);
      const [currentEvent, rawArguments] = queue.shift();
      const handler = entity.handlers.get(currentEvent);
      if (!handler) {
        let signature = entity.emittedEventTypes.get(currentEvent);
        if (!signature) {
          if (rawArguments.length !== 0) {
            throw new TevScriptError(
              "TEVS_IR_V3_EVENT_SIGNATURE_UNKNOWN",
              `unhandled event ${JSON.stringify(currentEvent)} has arguments but no portable signature`,
            );
          }
          signature = [];
        }
        if (signature.length !== rawArguments.length) {
          throw new TevScriptError("TEVS_IR_V3_EVENT_ARITY", `event ${JSON.stringify(currentEvent)} expects ${signature.length} arguments`);
        }
        const coerced = signature.map((typeId, index) =>
          coerceRuntimeV3(typeId, rawArguments[index], this.typeTable, `invoke ${entityId}.${currentEvent}[${index}]`));
        this.emitted.push({
          entity_id: entityId,
          event_id: currentEvent,
          argument_types: [...signature],
          arguments: coerced,
        });
        continue;
      }
      if (handler.parameters.length !== rawArguments.length) {
        throw new TevScriptError("TEVS_IR_V3_EVENT_ARITY", `event ${JSON.stringify(currentEvent)} expects ${handler.parameters.length} arguments`);
      }
      const parameters = new Map();
      handler.parameters.forEach((parameter, index) => {
        parameters.set(
          parameter.name,
          coerceRuntimeV3(parameter.type, rawArguments[index], this.typeTable, `invoke ${entityId}.${currentEvent}[${index}]`),
        );
      });
      const generated = this.executeHandler(entity, handler, parameters);
      for (const item of generated) {
        const event = {
          entity_id: entityId,
          event_id: item.eventId,
          argument_types: [...item.argumentTypes],
          arguments: item.args,
        };
        this.emitted.push(event);
        if (entity.handlers.has(item.eventId)) queue.push([item.eventId, item.args]);
      }
    }
    return this.emitted.slice(start);
  }

  executeHandler(entity, handler, parameters) {
    const instructions = handler.instructions;
    const budget = Number(handler.instruction_budget);
    const stack = [];
    const locals = new Map();
    const emitted = [];
    let pc = 0;
    let executed = 0;

    while (pc < instructions.length) {
      executed += 1;
      if (executed > budget) {
        throw new TevScriptError(
          "TEVS_IR_V3_INSTRUCTION_BUDGET",
          `handler ${JSON.stringify(handler.event_id)} exceeded instruction budget ${budget}`,
        );
      }
      const instruction = instructions[pc];
      switch (instruction.op) {
        case "CONST":
          stack.push(decodeV3Value(instruction.type, instruction.value, this.typeTable, {
            context: `instruction ${handler.event_id}:${pc}`,
          }));
          break;
        case "LOAD_STATE": stack.push(entity.state.get(instruction.name)); break;
        case "STORE_STATE": entity.state.set(instruction.name, stack.pop()); break;
        case "LOAD_LOCAL": stack.push(locals.get(instruction.name)); break;
        case "STORE_LOCAL": locals.set(instruction.name, stack.pop()); break;
        case "LOAD_PARAM": stack.push(parameters.get(instruction.name)); break;
        case "CONVERT_INT_TO_RAT": stack.push(new Rational(stack.pop())); break;
        case "UNARY": stack.push(unary(instruction.operator, stack.pop())); break;
        case "BINARY": {
          const right = stack.pop(); const left = stack.pop();
          stack.push(binaryV3(
            instruction.operator,
            instruction.left_type,
            instruction.right_type,
            left,
            right,
            this.typeTable,
          ));
          break;
        }
        case "CALL_PURE": {
          const arguments_ = popArguments(stack, Number(instruction.argc));
          const result = callPure(instruction.function_id, arguments_);
          if (instruction.return_type !== "Unit") stack.push(result);
          break;
        }
        case "CALL_CAPABILITY": {
          const capability = this.capabilities[instruction.capability_id];
          if (typeof capability !== "function") {
            throw new TevScriptError("TEVS_IR_V3_CAPABILITY_MISSING", `capability ${JSON.stringify(instruction.capability_id)} is not bound`);
          }
          const arguments_ = popArguments(stack, Number(instruction.argc));
          const result = capability(...arguments_);
          if (instruction.return_type !== "Unit") {
            stack.push(coerceRuntimeV3(
              instruction.return_type,
              result,
              this.typeTable,
              `capability ${instruction.capability_id} return`,
            ));
          }
          break;
        }
        case "EMIT_EVENT": {
          const argumentTypes = [...instruction.argument_types];
          const values = popArguments(stack, Number(instruction.argc));
          const arguments_ = argumentTypes.map((typeId, index) =>
            coerceRuntimeV3(typeId, values[index], this.typeTable, `emit ${instruction.event_id}[${index}]`));
          emitted.push({ eventId: instruction.event_id, argumentTypes, args: arguments_ });
          break;
        }
        case "MAKE_RECORD": {
          const descriptor = this.typeTable.require(instruction.type);
          const values = popArguments(stack, instruction.fields.length);
          const byName = new Map(instruction.fields.map((name, index) => [name, values[index]]));
          const record = new RecordValueV3(
            instruction.type,
            descriptor.fields.map(([name]) => [name, byName.get(name)]),
          );
          stack.push(decodeV3Value(
            instruction.type,
            encodeV3Value(instruction.type, record, this.typeTable),
            this.typeTable,
          ));
          break;
        }
        case "LOAD_FIELD": {
          const record = stack.pop();
          if (!(record instanceof RecordValueV3)) {
            throw new TevScriptError("TEVS_IR_V3_RUNTIME_RECORD", "LOAD_FIELD received non-record value");
          }
          stack.push(record.field(instruction.field));
          break;
        }
        case "MAKE_VARIANT": {
          const value = instruction.argc === 0
            ? new VariantValueV3(instruction.type, instruction.variant)
            : new VariantValueV3(instruction.type, instruction.variant, stack.pop());
          stack.push(decodeV3Value(
            instruction.type,
            encodeV3Value(instruction.type, value, this.typeTable),
            this.typeTable,
          ));
          break;
        }
        case "TEST_VARIANT": {
          const value = stack.pop();
          if (!(value instanceof VariantValueV3)) {
            throw new TevScriptError("TEVS_IR_V3_RUNTIME_VARIANT", "TEST_VARIANT received non-variant value");
          }
          stack.push(value.variant === instruction.variant);
          break;
        }
        case "LOAD_VARIANT_PAYLOAD": {
          const value = stack.pop();
          if (!(value instanceof VariantValueV3)
              || value.variant !== instruction.variant
              || !value.hasPayload) {
            throw new TevScriptError(
              "TEVS_IR_V3_VARIANT_UNWRAP",
              `cannot load payload for variant ${JSON.stringify(instruction.variant)}`,
            );
          }
          stack.push(value.payload);
          break;
        }
        case "JUMP_IF_FALSE": {
          const condition = stack.pop();
          if (condition !== true) { pc = Number(instruction.target); continue; }
          break;
        }
        case "JUMP": pc = Number(instruction.target); continue;
        case "RETURN": pc = instructions.length; continue;
        default: throw new TevScriptError("TEVS_IR_V3_OPCODE", `unknown V3 opcode ${JSON.stringify(instruction.op)}`);
      }
      pc += 1;
    }
    if (stack.length !== 0) {
      throw new TevScriptError(
        "TEVS_IR_V3_RUNTIME_STACK_LEAK",
        `handler ${JSON.stringify(handler.event_id)} left ${stack.length} values on stack`,
      );
    }
    return emitted;
  }

  state(entityId) {
    const entity = this.entities.get(entityId);
    if (!entity) throw new TevScriptError("TEVS_IR_V3_ENTITY_UNKNOWN", `unknown entity ${JSON.stringify(entityId)}`);
    return Object.fromEntries(entity.state);
  }

  canonicalState(entityId) {
    const entity = this.entities.get(entityId);
    if (!entity) throw new TevScriptError("TEVS_IR_V3_ENTITY_UNKNOWN", `unknown entity ${JSON.stringify(entityId)}`);
    return Object.fromEntries(
      [...entity.state.keys()].sort().map((name) => [
        name,
        encodeV3Value(entity.stateTypes.get(name), entity.state.get(name), this.typeTable, {
          context: `state ${entityId}.${name}`,
        }),
      ]),
    );
  }

  canonicalEventArguments(event) {
    return event.argument_types.map((typeId, index) =>
      encodeV3Value(typeId, event.arguments[index], this.typeTable, { context: `event ${event.event_id}[${index}]` }));
  }
}

export { validateProgramIrV3 } from "./ir-v3-validation.mjs";
export {
  RecordValueV3,
  VariantValueV3,
  Rational,
  TevScriptError,
  encodeV3Value,
} from "./ir-v3-values.mjs";
